### DISCUSSION TITLE
AuthApiError: invalid flow state, no valid flow state found

### DISCUSSION THREAD

**Discussion by kei2e** [supabase/supabase#16743] [Questions]
I was trying to implement Proof Key for Code Exchange (PKCE) in my authentication flow using Supabase Auth Helpers, and I kept receiving this error. So, I decided to see if I can make "User Management App with NextJS" example--from [Supabase Docs](https://supabase.com/docs/guides/getting-started/tutorials/with-nextjs)--works.

I still get the same error:

```
- error AuthApiError: invalid flow state, no valid flow state found
    at handleError (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/main/lib/fetch.js:36:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async _handleRequest (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/main/lib/fetch.js:84:9)
    at async _request (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/main/lib/fetch.js:63:18)
    at async SupabaseAuthClient._exchangeCodeForSession (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/main/GoTrueClient.js:377:33)
    at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/main/GoTrueClient.js:784:28) {
  __isAuthError: true,
  status: 403
}
``` 

I have also noticed that, before this is triggered, I get the following error:

```
TypeError: cookies is not iterable
    at get [headers map sorted] (node:internal/deps/undici/undici:1881:34)
    at node:internal/deps/undici/undici:1911:25
    at Headers Iterator.next (node:internal/deps/undici/undici:1121:26)
    at toNodeOutgoingHttpHeaders (./supabase-nextjs/node_modules/next/dist/server/web/utils.js:106:31)
    at invokeRender (./supabase-nextjs/node_modules/next/dist/server/lib/router-server.js:275:122)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async handleRequest (./supabase-nextjs/node_modules/next/dist/server/lib/router-server.js:447:24)
    at async requestHandler (./supabase-nextjs/node_modules/next/dist/server/lib/router-server.js:464:13)
    at async Server.<anonymous> (./supabase-nextjs/node_modules/next/dist/server/lib/start-server.js:117:13)
``` 

I don't seem to be able to find any solution to this. Does anyone know what the problem might be?

**Comment by Bartek532** [comment]
I'm experiencing exactly the same, which package are you using?

**Comment by N00nDay** [comment]
I am using SvelteKit / Supabase and this is happening to me when trying to use Google OAuth locally.

**Reply by shawnesquivel** [reply]
I'm experiencing the same with NextJS and Supabase. I can get the URL from my google auth

```
      const googleRes = await supabase.auth.signInWithOAuth({
        provider: "google",
        options: {
          redirectTo: "http://localhost:3000",
        },
      });
```


which returns  a URL with the code_challenge

```
{\"data\":{\"provider\":\"google\",\"url\":\"https://qctnbkfmquxoifgzwmml.supabase.co/auth/v1/authorize?provider=google&code_challenge=<CODE_CHALLENGE>&code_challenge_method=s256\"}
```

however when I pass it here, I get the same error as you all:
```
        const { data, error } = await supabase.auth.exchangeCodeForSession(
          code
        );

```

Error:
```
\"name\":\"AuthApiError\",
\"message\":\"invalid flow state, no valid flow state found\",\
"status\":403}"

```

**Reply by diego-yason** [reply]
Have you guys figured this out? Got stuck on the exact same error when trying to use Supabase locally.
@N00nDay @shawnesquivel

**Reply by SonnyFishback** [reply]
I was not able to find any documentation anywhere to explain this well so I hope this helps. @diego-yason.

In my case I ran into this issue while implementing Google OAuth w/ SvelteKit  w/ Local Supabse however, I think the same concept translates across OAuth Clients/ Metaframeworks.

1. You must configure your OAuth Client to redirect to  API URL  `http://127.0.0.1:54321` of your local Supabase instance 

```
Authorized JavaScript origins
For use with requests from a browser
URIs 1 
https://******************.vercel.app
URIs 2 
http://localhost:5173
URIs 3 
http://localhost:54321
URIs 4 
http://127.0.0.1:54321
```

```
Authorized redirect URIs
URIs 1 
https://${PUBLIC_SUPABASE_URL}/auth/v1/callback
URIs 2 
http://localhost:5173/auth/v1/callback
URIs 3 
http://localhost:54321/auth/v1/callback
URIs 4 
http://127.0.0.1:54321/auth/v1/callback
```

2. You must configure your `config.toml` correctly to allow redirects to the proper endpoint. Here is mine:

```
[auth]
enabled = true
# The base URL of your website. Used as an allow-list for redirects and for constructing URLs used
# in emails.
site_url = "http://localhost:5173/"
# A list of *exact* URLs that auth providers are permitted to redirect to post authentication.
additional_redirect_urls = ["https://127.0.0.1:3000"]

......... MORE CONFIG

[auth.external.google]
enabled = true
client_id = "env(SUPABASE_AUTH_EXTERNAL_GOOGLE_CLIENT_ID)"
secret = "env(SUPABASE_AUTH_EXTERNAL_GOOGLE_SECRET)"
# Overrides the default auth redirectUrl.
redirect_uri = "http://127.0.0.1:54321/auth/v1/callback"
# Overrides the default auth provider URL. Used to support self-hosted gitlab, single-tenant Azure,
# or any other third-party OIDC providers.
url = ""
```

**Reply by ShriPunta** [reply]
Thank you for this message. It took me 2 days of debugging and I was going nowhere. The supabase documentation was not clear at all.

Hope other people find it helpful.

**Reply by benwestdev** [reply]
Interesting...I am building in a sveltekit app as well and when I followed your advice I get this error: {"message":"no Route matched with those values"}
Anyone else get this? 

Edit: Figured it out! Had a typo in redirect_uri. Make sure you literally use http://127.0.0.1:54321/auth/v1/callback

**Reply by wilsonam** [reply]
@SonnyFishback thank you so much for this! I ran into the same issue with Supabase and Nextjs locally. I've been searching docs for days before I found this thread.

**Reply by hwangio** [reply]
Many thanks for @SonnyFishback's solution! Following this direction here is a simpler solution that works for me:
1. (Google Cloud Console) keep 'Authorized JavaScript origins' unchanged (e.g., http://localhost:3000)
2. ⭐️ (Google Cloud Console)  add to 'Authorized redirect URIs' the local supabase api, e.g., http://localhost:54321/auth/v1/callback
3. ⭐️ for `supabase/config.toml`, change `[auth.external.google]` section according to @SonnyFishback 

My understanding: we need for development a local version of the "https://some-project-id.supabase.co/auth/v1/callback", i.e., http://localhost:54321/auth/v1/callback -- this is for Google to communicate with supabase auth service, so it should be included in GC console settings.
The other callback url, e.g., http://localhost:3000/api/auth/callback, is used downstream, inside `signInWithOAuth` function. It's used by Supabase Auth service to send & deal with google auth code. So this api should be included in supabase (local) config.toml's `[auth]` section, i.e., site_url / additional_redirect_urls (or at least according to doc; in my case it works even if it's not).

**Comment by dijonmusters** [comment]
The `redirectTo` option needs to redirect to a Code Exchange route. I added a deeper explanation [here](https://github.com/supabase/auth-helpers/issues/718#issuecomment-1895499884) but just realised the Next.js starter suggestion is probably not very helpful if you're building a SvelteKit app 😆

Hopefully [this one](https://github.com/supabase-community/supabase-by-example/blob/main/oauth-flow/sveltekit/src/routes/(auth)/auth/callback/%2Bserver.ts) can steer you in the right direction for declaring a SvelteKit route that can handle the code exchange 👍

**Reply by asnimansari** [reply]
Can't find this page.

**Comment by Zizouducode** [comment]
Here is my config : 

Front:  NextJS
Backend: NestJS with Supabase

I followed this [tutorial](https://supabase.com/docs/guides/auth/social-login/auth-google?queryGroups=environment&environment=server#application-code) with the server side connection. 

On my side the issue was in my config.toml. 

I added the redirect_url here  in my config.toml file: additional_redirect_urls = ["https://localhost:3000", "http://localhost:3000/auth/callback"]

I fixed the issue when I commented this line in the google auth provider : #redirect_uri = "" in my config.toml. I was passing the URI from my .env file. But it seems there is a mismatch between this URI and the one sent in the frontend by the supabase method.

**Comment by arjunbhut** [comment]
Thanks , this discussion is very helpful, documentations is not very clear, need to go across github, reddit and other sources to find the solutions

**Comment by MuellerHF** [comment]
I use docker with compose file for selfhosted supabase. I ran into the same issue and tried out a bunch of things until I got it running. Here is what I did:

Setup OAuth for external Auth:
I had to edit my compose file. To be specific supabase-auth:
    image: 'supabase/gotrue'

Under enviroment first add the envs you need https://github.com/supabase/auth/blob/master/example.env

Important here is following:
`GOTRUE_EXTERNAL_DISCORD_REDIRECT_URI=https://<db.yourdomain.com>/auth/v1/callback`

For example if you check this docs:
https://supabase.com/docs/guides/auth/social-login/auth-discord
It literally tells you to set up this url as redirect url on discord.
The Diamond Brackets part is the url where your supabase is hosted and running. I usually prefer "db" as subdomain.

If you want to handle the code you obivously have to create an endpoint in your app and set it as redirect:

e.g.:

`const { data, error } = await supabase.auth.signInWithOAuth({
				provider: provider,
				options: {
					redirectTo: `${window.location.origin}/auth/callback`
				}
			});`

And then use the exchangeCodeForSession method on this server enpoint. This should do the trick.

I think the flow is like this:

1. You press login button triggering the signInWithOAuth
2. The Client will be redirected to the provider and confirms action
3. The provider redirects to your supabase auth/v1/callback endpoint
4. Supabase verifies and sends the user back to the given redirectTo including the exchangable code
5. You handle the code and login the user safely on the server

My fault was I was trying to let discord directly redirect back to my page instead of the db server auth endpoint. But the supabase endpoint will trigger the redirect to my page after confirming the action. Hope this helps someone
